### ISSUE TITLE
AssimpImporter texture mapping (obj/mtl)

### ISSUE THREAD

**Issue by IcySnex** [RolandKoenig/SeeingSharp2#12]
Is this library able to map textures imported by AssimpImporter?
I can't seem to load a model in obj format which has a correctly mapped texture (png).
I already tried multiple different models and file formats but even the sample project (ModelViewer) doesnt respect my textures inside the .mtl file.
I notice that the mtl correctly gets loaded into the render resources (as the #Resources counter goes one up when including the material), but still no image (only the gray model).
‎ 
This is what my mtl file looks like:
```
newmtl TextureMaterial
map_Kd skin.png
```
And im loading my model with this code:
```cs
loadedModel = await GraphicsCore.Current.ImportersAndExporters.ImportAsync(modelLink, importOptions);
await renderLoop.Scene.ImportAsync(loadedModel);
```
‎ 
Its no problem to texture some custom meshes but how would you texture an entire model?
This is the code I am using to texture a simple cube:
```cs
await renderLoop.Scene.ManipulateSceneAsync(manipulator =>
{
    var resTexture = manipulator.AddTextureResource(LoadedTexure);
    var resMaterial = manipulator.AddStandardMaterialResource(resTexture);
    var resGeometry = manipulator.AddResource(d => new GeometryResource(new CubeGeometryFactory()));

    var mesh = new Mesh(resGeometry, resMaterial);
    manipulator.AddObject(mesh);
});
```
‎ 
Im sorry if I am missing something or if this is a useless question but Im kinda stuck on this.
Also a quick note: This project is amazing! It runs smooth, has animation support and is quite easy to use. Thank you for your effort.

**Comment by RolandKoenig** [comment]
You are right, the assimp importer of SeeingSharp just ignores texture mapping by now
I can look into this in the next days

**Comment by IcySnex** [comment]
Thank you for the very fast awnser and solution.
So loading textures now fully works.
But I still ran into some specific use case issues just for my model/texture (not in general).
I have fixed some parts myself but this created some other problems instead.
‎

My model is a rather simple character model for a game. It only contains 12 cubes/geometries and has a transparent 64x64 texture. I have uploaded the model/texture here: [dropbox,.com](https://www.dropbox.com/sh/3pr2so4tmixtymv/AACOw5uD9D3pSk3sj-ANvUKSa?dl=0).
This is what my model looks like in Blender and other 3D rendering softwares:
![blender](https://user-images.githubusercontent.com/81762750/199700127-570b98f8-6231-4d3a-82c7-19cc0c781d2a.png) ![blenderBack](https://user-images.githubusercontent.com/81762750/199701940-764540b4-394e-46e1-93b5-4a1f0feb018c.png)

**Issue 1: Transparent Textures**
After loading my model without changing anything, it looked like this:
![first](https://user-images.githubusercontent.com/81762750/199699651-1659c7be-d0dd-488d-8711-4734965b15c0.png)
As you can see the transparent parts of the image are not rendered transparent, instead the default diffusion color is rendered.
This can be fixed pretty easily just by setting ```materialResource.ClipFactor = 1f;``` and ```materialResource.UseVertexColors = true``` inside the ```AssimpImporter.ProcessMaterials(...)``` method.
Of course, applying these changes results in completely invisible models when there is no texture set. But this is just for debug testing at the moment so that a correct solution can be implemented later.

After these changes the model looks like this:
![clipFront](https://user-images.githubusercontent.com/81762750/199701306-d80f076d-e33d-448e-838d-769511f674ef.png) ![clipBack](https://user-images.githubusercontent.com/81762750/199701541-a1f18b98-06e6-47a2-8629-947b30b1abe7.png)
Definitely better, right? But now we have some weird artifacts at the borders. This is espacially visible from the back.
I tried to change some settings on the material but I coudlnt get it to work correctly.
‎ 
‎ 
‎**Issue 2: Upscaling Textures**
So my next issue is that the texture is scaled up linearly instead of neighrest neighbar, which results in a blurry texture.
There was no build in parameter to set the interpolation mode so I tried to set a ```BitmapResource``` with a ```Direct2DOneTimeRenderTextureResource``` as the texture manually inside the ```AssimpImporter.ProcessTextures(...)``` method.
If you are drawing the Bitmap yourself, there is an option to set the BitmapInterpolationMode.
So I replaced this code:
```cs
modelContainer.AddResource(new ImportedResourceInfo(
    modelContainer.GetResourceKey("Texture", actMaterial.TextureDiffuse.FilePath),
    _ => new StandardTextureResource(
        currentFileLink.GetForAnotherFile(actMaterial.TextureDiffuse.FilePath))));
```
with this code:
```cs
modelContainer.AddResource(new ImportedResourceInfo(
    modelContainer.GetResourceKey("Texture", actMaterial.TextureDiffuse.FilePath),
    _ =>
    {
        var d2DDrawingLayer = new Custom2DDrawingLayer(graphics =>
        {
            var d2DRectangle = new RectangleF(0, 0, 64, 64);
            var bitmap = new StandardBitmapResource(currentFileLink.GetForAnotherFile(actMaterial.TextureDiffuse.FilePath));
            graphics.DrawBitmap(bitmap, d2DRectangle, 1, BitmapInterpolationMode.NearestNeighbor);
        });
        return new Direct2DOneTimeRenderTextureResource(d2DDrawingLayer, 64, 64);
    }));
```
Of course, these changes only apply to this texture, since the height and width are hard-coded in the render function. A real solution could be implemented at a later date.
But even after changing this there was no effect on the end result. The texture remained blurred.

Sorry if I'm missing something completely, I'm fairly new to 3D rendering in general.
Thank you for your great help and have a nice day!

**Comment by RolandKoenig** [comment]
Thanks for providing me the model, this helps to find the way for a clean solution. I go into this more deeper in the next days

Btw: An interesting idea to use a Custom2DDrawingLayer here. 
Try to not load a resource inside it, because it gets called in each frame. You can create the StandardBitmapResource outside of the scope

**Comment by RolandKoenig** [comment]
A short status update:
I tried some things this evening and figured out the points where SeeingSharp needs a little improvement. See the current result below:

![04-11-_2022_22-29-15](https://user-images.githubusercontent.com/11254388/200077823-1c109933-848a-40d9-975e-93fc8ba6b1f6.png)

The things I need to add now are the following
 - Enable to configure texture filters on material level (or texture level) --> Currently SeeingSharp only uses anisotropic filters. This results in the blurry textures from your screenshots.
 - The way SeeingSharp processes transparent texels from textures currently is wrong for your scenario. By default, SeeingSharp mixes the texture and vertex colors based on the alpha value. In your case those have to be clipped away
 
 to sum up... It should not be hard to modify SeeingSharp in that way that your scenario can be achieved very easily

**Comment by IcySnex** [comment]
Wow, this looks amazing!
Cant wait to use it in my application soon.
Thank you very much for helping me with this.

**Comment by RolandKoenig** [comment]
I've updated the branch https://github.com/RolandKoenig/SeeingSharp2/tree/feature/assimp_import_textures
Now all looks pretty good. I do only want to do some further testing before it gets merged into the main branch

You can achieve now almost all you want by standard except a short configuration for the importer. So far as I see I get no hint from assimp that the texture should be displayed pixelated. Therefore SeeingSharp uses the default texture filter which is anisotropic and produces the blurry image. You can change this behavior by passing AssimpImportOptions to the importer and update the texture filter like below. The filter SeeingSharpFilter.MinMagMipPoint produces the pixelated image just like in the sceenshots above. 

`var assimpImportOptions = new AssimpImportOptions();
assimpImportOptions.ConfigureTextureAction = (filePath, texture) =>
{
    texture.Filter = SeeingSharpFilter.MinMagMipPoint;
};

// Import`

I also created a new sample in the SampleContainer called "TextureSampling". There you can play around with different texture filters

![image](https://user-images.githubusercontent.com/11254388/200124424-a0353495-f833-491e-8be3-c10e187386e7.png)

**Comment by IcySnex** [comment]
I just tested it and it works great with the model now.
Here is a screenshot:
![done](https://user-images.githubusercontent.com/81762750/200164249-2c5fc585-8609-42cd-9e7f-5399c20d1e37.png)

I will close this issue now, thank you.
